EOS 3.9. Hack extension for gnome 3.38

danigm opened the pull request:
https://phabricator.endlessm.com/T30449

GeorgesStavracas commented on utils.js:
```diff
@@ -178,7 +179,11 @@ let _currentDesktopsMatches = {};
+function is(name, maxVersion = '3.36') {
+    if (config.PACKAGE_VERSION > maxVersion) {
+        return false;
+    }
```
Nit: don't need curly braces

GeorgesStavracas commented on utils.js:
```diff
@@ -165,3 +165,38 @@ class ObjectsMap {
     }
 };
+
+let _currentDesktopsMatches = {};
```
This can be `const` I guess

GeorgesStavracas replied:
(I know this is copy pasted code, but I didn't want to change the original commit since it's not made by us)

GeorgesStavracas commented on utils.js:
```diff
@@ -165,3 +165,38 @@ class ObjectsMap {
+//
+// This function is a copy of:
+// https://github.com/endlessm/gnome-shell/blob/master/js/misc/desktop.js
+function is(name) {
```
The original commit has the benefit of allowing us to write `Desktop.is()`, but this nuance is lost here. Perhaps it's better to rename this function to `desktopIs()` so that we can call it as `Util.desktopIs('endless')`. What do you think?

GeorgesStavracas commented on utils.js:
```diff
@@ -165,3 +165,38 @@ class ObjectsMap {
+        return _currentDesktopsMatches[name];
+    }
+
+    let desktopsEnv = GLib.getenv('XDG_CURRENT_DESKTOP');
```
`const`

GeorgesStavracas commented on utils.js:
```diff
@@ -165,3 +165,38 @@ class ObjectsMap {
+        return false;
+    }
+
+    let desktops = desktopsEnv.split(":");
```
This can be simplified as:

```javascript
const hasMatch = desktopEnv.split(':').some(desktop => desktop === name);
_currentDesktopsMatches[name] = hasMatch;
return hasMatch;
```

danigm replied:
Done, I've replaced the code with this one.

GeorgesStavracas commented on service.js:
```diff
@@ -300,6 +300,10 @@ function enable() {
 
+    if (!Utils.is('endless')) {
+        return;
+    }
```
Nit: don't need curly braces for single-line ifs

GeorgesStavracas commented on service.js:
```diff
@@ -335,7 +339,9 @@ function enable() {
-    Utils.restore(ShellDBus.AppStoreService);
+    if (Utils.is('endless')) {
+        Utils.restore(ShellDBus.AppStoreService);
+    }
```
Nit: don't need curly braces for single-line ifs

GeorgesStavracas commented on ui/codeView.js:
```diff
@@ -89,8 +92,67 @@ function _getWindowId(win) {
+    const targetPropertyVariant = proxy.get_cached_property('Target');
+    if (!targetPropertyVariant) {
+        return null;
+    }
```
Nit: don't need curly braces for single-line ifs

GeorgesStavracas commented on ui/codeView.js:
```diff
@@ -89,8 +92,67 @@ function _getWindowId(win) {
+                                       windowAppId,
+                                       windowObjectPath,
+                                       hackToolboxId,
+                                       null);
```
The modern code style is:

```javascript
proxy = Gio.DBusProxy.new_sync(
    Gio.DBus.session,
    Gio.DBusProxyFlags.DO_NOT_AUTO_START |
    Gio.DBusProxyFlags.DO_NOT_CONNECT_SIGNALS,
    null,
    windowAppId,
    windowObjectPath,
    hackToolboxId,
    null);
```

GeorgesStavracas commented on ui/codeView.js:
```diff
@@ -89,8 +92,67 @@ function _getWindowId(win) {
+     * com.hack_computer.HackToolbox */
+    if (windowAppId.startsWith('com.endlessm')) {
+        hackToolboxId = OLD_HACK_TOOLBOX_ID;
+    }
```
Nit: don't need curly braces for single-line ifs

GeorgesStavracas commented on ui/codeView.js:
```diff
@@ -89,8 +92,67 @@ function _getWindowId(win) {
+    /* Not a bus name, no way that this could be a toolbox */
+    if (!Gio.dbus_is_name(windowAppId)) {
+        return null;
+    }
```
Nit: don't need curly braces for single-line ifs

GeorgesStavracas commented on ui/codeView.js:
```diff
@@ -89,8 +92,67 @@ function _getWindowId(win) {
+
+    if (!windowAppId || !windowObjectPath) {
+        return null;
+    }
```
Nit: don't need curly braces for single-line ifs

GeorgesStavracas commented on ui/codeView.js:
```diff
@@ -89,8 +92,67 @@ function _getWindowId(win) {
+    // This will work only on EndlessOS
+    if (Shell.WindowTracker.get_hack_toolbox_proxy) {
+        return Shell.WindowTracker.get_hack_toolbox_proxy(win);
+    }
```
Is it possible to not rely on `get_hack_toolbox_proxy` anymore? If we can drop the patch that adds the hack toolbox, it'll be better.

danigm replied:
Yes, I think so. This patch provides a fallback behavior for this so I think we drop that patch. There are some issues yet, with windows pairing (app + toolbox), but I'll fix that overriding some functions on the dash icon and overview mode.

GeorgesStavracas commented on ui/codeView.js:
```diff
@@ -1730,11 +1730,19 @@ function getWindows(workspace) {
     }).filter((w, i, a) => !w._hackIsInactiveWindow && !w.skip_taskbar && a.indexOf(w) === i);
 }
 
+function is_speedwagon_window(metaWindow) {
```
I guess we can drop this SpeedWagon thing entirely. We're dropping SpeedWagon from the desktop.

danigm replied:
I've removed all the speedwagon related code.

GeorgesStavracas commented on ui/wobblyFx.js:
```diff
@@ -0,0 +1,609 @@
+        changed: function () {},
+        disconnect: function () {},
+    };
+};
```
Since we're already modifying the extension, should we also remove Prefs?

danigm replied:
I've removed the Prefs object.

GeorgesStavracas commented on ui/codeView.js:
```diff
@@ -1761,7 +1772,72 @@ function getWindows(workspace) {
+    // If the app is flipped we should activate the toolbox window
+    if (session && session.flipped) {
+        win = session.toolbox.meta_window;
+    }
```
Nit: don't need curly braces

GeorgesStavracas commented on ui/codeView.js:
```diff
@@ -1761,7 +1772,72 @@ function getWindows(workspace) {
+    const wins = allWindows.filter(w => {
+        if (toolboxSession && toolboxSession.toolbox.meta_window === w) {
+            return false;
+        }
```
Nit: don't need curly braces

GeorgesStavracas commented on ui/codeView.js:
```diff
@@ -1761,7 +1772,72 @@ function getWindows(workspace) {
+
+        if (appSession && appSession.toolbox && appSession.toolbox.meta_window === w) {
+            return true;
+        }
```
Nit: don't need curly braces

GeorgesStavracas commented on ui/codeView.js:
```diff
@@ -1761,7 +1772,72 @@ function getWindows(workspace) {
+
+        if (appSession && appSession.flipped && appSession.app.meta_window === w) {
+            return false;
+        }
```
Nit: don't need curly braces

GeorgesStavracas commented on ui/codeView.js:
```diff
@@ -1859,6 +1859,26 @@ function addButton(app) {
+        const focusApp = windowTracker.focus_app;
+        if (focusApp.get_id().slice(0, -8) === 'com.hack_computer.Clubhouse') {
+            Main.panel.statusArea['appIcons']._setActiveApp(focusApp);
+        }
```
Nit: don't need curly braces

GeorgesStavracas commented on ui/appDisplay.js:
```diff
@@ -131,7 +131,18 @@ class HackAppIcon extends AppDisplay.AppIcon {
+                              'Activate', params, null,
+                              Gio.DBusCallFlags.NONE,
+                              -1, null,
+                              (conn, res) => conn.call_finish(res));
```
Nit: new style (as previously commented)

danigm commented:
Thanks for the review. I've passed eslint and fixed a lot of issues.

I've also applied the suggestions and fixed some problems in the wobbly windows effect.

And about the `get_hack_toolbox_proxy`, the shell code pair the toolbox window with the "app", so that behavior needs to be replaced on each specific place. Without that the extension should tweak all the places where the window-shellApp relation is important, like the alt-tab, overview or dash icons.

GeorgesStavracas commented on ui/codeView.js:
```diff
@@ -1208,6 +1209,12 @@ var CodingSession = GObject.registerClass({
+            this._attachBackendWindow();
+        } else {
+            this._detachBackendWindow();
+        }
```
Nit: don't need any of the curly braces
